### PR TITLE
feature: support password decryption

### LINKED ISSUE #2665: file.conf配置数据库密码不支持加密
### Ⅰ. file.conf配置，当使用数据库的时候，好像不支持密码加密

mode = "db"的时候，使用mysql数据库，password希望能够支持对生产数据库的密码加密，现在只能是明文格式，会有密码泄露的风险，能有什么解决方法吗，或者预计在那个版本会加入这个功能？

### PR DESCRIPTION
**Pull request by a364176773**
### Ⅰ. Describe what this PR did
support druid encryption and decryption

### Ⅱ. Does this pull request fix one issue?

fix #2665

### CODE REVIEW

File: server/src/main/java/io/seata/server/store/DruidDataSourceGenerator.java [hunk 1]
```diff
@@ -29,13 +38,30 @@
...
     public DataSource generateDataSource() {
         DruidDataSource ds = new DruidDataSource();
         ds.setDriverClassName(getDriverClassName());
         ds.setDriverClassLoader(getDriverClassLoader());
         ds.setUrl(getUrl());
         ds.setUsername(getUser());
+        String publicKey = getPublicKey();
+        if (StringUtils.isNotBlank(publicKey)) {
+            try {
+                ds.setFilters("config");
```
Comment (slievrly) [inline server/src/main/java/io/seata/server/store/DruidDataSourceGenerator.java:54]: if the filter is necessary?

Comment (a364176773) [reply server/src/main/java/io/seata/server/store/DruidDataSourceGenerator.java:54]: > if the filter is necessary?

ok

File: core/src/main/java/io/seata/core/store/db/AbstractDataSourceProvider.java [hunk 2]
```diff
@@ -195,6 +202,18 @@ protected String getUser() {
      */
     protected String getPassword() {
         String password = CONFIG.getConfig(ConfigurationKeys.STORE_DB_PASSWORD);
+        String publicKey = getPublicKey();
+        if (StringUtils.isNotBlank(publicKey)) {
+            try {
+                password = ConfigTools.decrypt(publicKey, password);
+            } catch (Exception e) {
+                if (LOGGER.isErrorEnabled()) {
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/store/db/AbstractDataSourceProvider.java:210]: The error log does not need to add this judgment.

File: core/src/main/java/io/seata/core/constants/ConfigurationKeys.java [hunk 3]
```diff
@@ -229,6 +229,11 @@
      */
     public static final String STORE_DB_LOG_QUERY_LIMIT = STORE_DB_PREFIX + "queryLimit";
 
+    /**
+     * The constant STORE_DB_PUBLIC_KER.
+     */
+    public static final String STORE_DB_PUBLIC_KER = STORE_DB_PREFIX + "publicKey";
```
Comment (l81893521) [inline core/src/main/java/io/seata/core/constants/ConfigurationKeys.java:235]: KER or KEY?

File: core/src/main/java/io/seata/core/store/db/AbstractDataSourceProvider.java [hunk 4]
```diff
@@ -195,6 +202,18 @@ protected String getUser() {
      */
     protected String getPassword() {
         String password = CONFIG.getConfig(ConfigurationKeys.STORE_DB_PASSWORD);
+        String publicKey = getPublicKey();
+        if (StringUtils.isNotBlank(publicKey)) {
+            try {
+                password = ConfigTools.decrypt(publicKey, password);
```
Comment (l81893521) [inline core/src/main/java/io/seata/core/store/db/AbstractDataSourceProvider.java:208]: ConfigTools is the class of druid. Is any problem when use other? And this is the abstract class, I think you should override the method of 
```
getPassword()
```

File: server/src/main/resources/file.conf [hunk 5]
```diff
@@ -37,6 +37,7 @@ store {
     lockTable = "lock_table"
     queryLimit = 100
     maxWait = 5000
+    publicKey = ""
```
Comment (jsbxyyx) [inline server/src/main/resources/file.conf:40]: write a comment for `publicKey`

Comment (a364176773) [reply server/src/main/resources/file.conf:40]: ok

File: core/src/main/java/io/seata/core/store/db/AbstractDataSourceProvider.java (hunk 4 above)
Comment (a364176773) [reply core/src/main/java/io/seata/core/store/db/AbstractDataSourceProvider.java]: ok

File: core/src/main/java/io/seata/core/constants/ConfigurationKeys.java (hunk 3 above)
Comment (a364176773) [reply core/src/main/java/io/seata/core/constants/ConfigurationKeys.java]: ok

File: core/src/main/java/io/seata/core/store/db/AbstractDataSourceProvider.java (hunk 2 above)
Comment (a364176773) [reply core/src/main/java/io/seata/core/store/db/AbstractDataSourceProvider.java:207]: ok

File: common/src/main/java/io/seata/common/util/ConfigTools.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        String publicKeyStr = ConfigTools.getPublicKey(keyPair);
+        String privateKeyStr = ConfigTools.getPrivateKey(keyPair);
+        System.out.println("publicKeyStr:\n" + publicKeyStr);
+        System.out.println("privateKeyStr:\n" + privateKeyStr);
+        System.out.println(
+            "after the key is generated, please keep your key pair properly, if you need to encrypt, please enter your database password");
+        System.out.println("input 'q' exit");
+        while (scan.hasNextLine()) {
+            String password = scan.nextLine();
+            if (StringUtils.isNotBlank(password) && !password.equalsIgnoreCase("q")) {
```
Comment (jsbxyyx) [inline common/src/main/java/io/seata/common/util/ConfigTools.java:138]: password char has \n is right?

File: common/src/main/java/io/seata/common/util/ConfigTools.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+import java.security.KeyPairGenerator;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Scanner;
+
+import javax.crypto.Cipher;
+
+import sun.misc.BASE64Decoder;
```
Comment (jsbxyyx) [inline common/src/main/java/io/seata/common/util/ConfigTools.java:30]: use java.util.Base64

Comment (a364176773) [reply common/src/main/java/io/seata/common/util/ConfigTools.java:30]: ok

File: common/src/main/java/io/seata/common/util/ConfigTools.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Base64;
+import java.util.Scanner;
+import javax.crypto.Cipher;
+
+/**
+ * @author funkye
+ */
+public class ConfigTools {
```
Comment (caohdgege) [inline common/src/main/java/io/seata/common/util/ConfigTools.java:32]: 这个类里面的一些方法其实是不会抛出异常的，比如getKeyPair；除此之外，其他有跑异常的方法，感觉都可以转换成ShouldNeverHappenException
然后还有 "RSA", "UTF-8"感觉改成常量会好点 StandardCharsets.UTF_8

**Comment by caohdgege** [comment]
add publicKey to config.txt

File: common/src/main/java/io/seata/common/util/ConfigTools.java (hunk 8 above)
Comment (a364176773) [reply common/src/main/java/io/seata/common/util/ConfigTools.java:32]: ok

File: common/src/main/java/io/seata/common/util/ConfigTools.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    public static String byte2Base64(byte[] bytes) {
+        return new String(Base64.getEncoder().encode(bytes), StandardCharsets.UTF_8);
+    }
+
+    // Base64 encoding to byte array
+    public static byte[] base642Byte(String base64Key) {
+        return Base64.getDecoder().decode(base64Key);
+    }
+
+    public static void main(String[] args) throws Exception {
```
Comment (l81893521) [inline common/src/main/java/io/seata/common/util/ConfigTools.java:121]: please remove the main

**Review by caohdgege — APPROVED** [review]
LGTM

File: server/src/main/java/io/seata/server/storage/redis/JedisPooledFactory.java [hunk 10]
```diff
@@ -74,8 +77,19 @@ public static JedisPoolAbstract getJedisPoolInstance(JedisPoolAbstract... jedisP
                         jedisPool = jedisPools[0];
                     } else {
                         String password = CONFIGURATION.getConfig(ConfigurationKeys.STORE_REDIS_PASSWORD);
-                        if (StringUtils.isBlank(password)) {
+                        if (StringUtils.isBlank(password) || StringUtils.equalsIgnoreCase(password, DEFAULT_PASSWORD)) {
```
Comment (jsbxyyx) [inline server/src/main/java/io/seata/server/storage/redis/JedisPooledFactory.java:80]: password is null string. ignore?

Comment (a364176773) [reply server/src/main/java/io/seata/server/storage/redis/JedisPooledFactory.java]: done

**Review by jsbxyyx — APPROVED** [review]
LGTM

**Review by l81893521 — APPROVED** [review]
LGTM for @caohdgege

**Comment by quanqiuchaoren** [comment]
你们好，我看所有关于加解密的都会被引导到这个地方，但是我没有在这个地方找到如何进行加解密的方式，或许这个文章中提到了一些关键词，但是我对这个方面的知识比较少，目前还是不知道怎么加解密，如果有人知道的话，请告知具体的方式，感谢

**Comment by quanqiuchaoren** [comment]
Hello, I see that everything about encryption and decryption leads to this place, but I didn't find how to encrypt and decrypt in this place. Maybe there are some keywords mentioned in this article, but I have little knowledge about this. I still don't know how to encrypt and decrypt. If anyone knows, please tell me how. Thanks.

**Pull request merged**